src/hierarchies/mfa_hierarchy.h

```diff
@@ -0,0 +1,132 @@
+  bayesmix::HierarchyId get_id() const override {
+    return bayesmix::HierarchyId::MFA;
+  }
+
+  //! Read and set state values from a given Protobuf message
+  void set_state_from_proto(const google::protobuf::Message& state_) override;
+
+  //! Read and set hyperparameter values from a given Protobuf message
+  void set_hypers_from_proto(
+      const google::protobuf::Message& hypers_) override;
+
+  //! Writes current state to a Protobuf message and return a shared_ptr
+  //! New hierarchies have to first modify the field 'oneof val' in the
+  //! AlgoritmState::ClusterState message by adding the appropriate type
+  std::shared_ptr<bayesmix::AlgorithmState::ClusterState> get_state_proto()
+      const override;
+
+  //! Writes current value of hyperparameters to a Protobuf message and
+  //! return a shared_ptr.
+  //! New hierarchies have to first modify the field 'oneof val' in the
+  //! AlgoritmState::HierarchyHypers message by adding the appropriate type
+  std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers> get_hypers_proto()
+      const override;
+
+  //! Returns whether the hierarchy models multivariate data or not
+  bool is_multivariate() const override { return true; }
+
+  //! Saves posterior hyperparameters to the corresponding class member
+  void save_posterior_hypers() {
+    // TODO
+    /* posterior_hypers =
+        static_cast<Derived *>(this)->compute_posterior_hypers();*/
+  }
+
+  //! Generates new state values from the centering posterior distribution
+  //! @param update_params  Save posterior hypers after the computation?
+  void sample_full_cond(bool update_params = false) override;
+
+ protected:
+  //! Evaluates the log-likelihood of data in a single point
+  //! @param datum      Point which is to be evaluated
+  //! @return           The evaluation of the lpdf
+  double like_lpdf(const Eigen::RowVectorXd& datum) const override;
+
+  //! Updates cluster statistics when a datum is added or removed from it
+  //! @param datum      Data point which is being added or removed
+  //! @param add        Whether the datum is being added or removed
+  void update_summary_statistics(const Eigen::RowVectorXd& datum,
+                                 bool add) override;
+
+  //! Initializes state parameters to appropriate values
+  void initialize_state() override;
+
+  //! Initializes hierarchy hyperparameters to appropriate values
+  void initialize_hypers() override;
+
+  //! Gibbs sampling step for state variable eta
+  void sample_eta();
+
+  //! Gibbs sampling step for state variable mu
+  void sample_mu();
+
+  //! Gibbs sampling step for state variable psi
+  void sample_psi();
+
+  //! Gibbs sampling step for state variable lambda
+  void sample_lambda();
+
+  //! Sum of data points currently belonging to the cluster
+  Eigen::VectorXd data_sum;
+
+  //! Number of variables for each datum
+  size_t dim;
+
+  //! Vector of data points currently belonging to the cluster
+  std::vector<Eigen::VectorXd> data;
```

Comment:
In the `BaseHierarchy`, we already save `std::set<int> cluster_data_idx` where we collect the indexes of all the data-points in the cluster.
Instead of saving in addition an `std::vector` with the data points, you could store a pointer to the full dataset, and instead of writing `data[I]` you could write `all_observations[cluster_data_idx[i]]`. (all_observations could be equivalently called data)